src/ledger/LedgerTxnOfferSQL.cpp

```diff
@@ -65,27 +65,81 @@ LedgerTxnRoot::Impl::loadAllOffers() const
 std::deque<LedgerEntry>::const_iterator
 LedgerTxnRoot::Impl::loadBestOffers(std::deque<LedgerEntry>& offers,
                                     Asset const& buying, Asset const& selling,
-                                    size_t numOffers, size_t offset) const
+                                    size_t numOffers) const
 {
+    // price is an approximation of the actual n/d (truncated math, 15 digits)
+    // ordering by offerid gives precendence to older offers for fairness
     std::string sql = "SELECT sellerid, offerid, sellingasset, buyingasset, "
                       "amount, pricen, priced, flags, lastmodified "
                       "FROM offers "
-                      "WHERE sellingasset = :v1 AND buyingasset = :v2";
+                      "WHERE sellingasset = :v1 AND buyingasset = :v2 "
+                      "ORDER BY price, offerid LIMIT :n";
 
     std::string buyingAsset, sellingAsset;
     buyingAsset = decoder::encode_b64(xdr::xdr_to_opaque(buying));
     sellingAsset = decoder::encode_b64(xdr::xdr_to_opaque(selling));
 
+    auto prep = mDatabase.getPreparedStatement(sql);
+    auto& st = prep.statement();
+    st.exchange(soci::use(sellingAsset));
+    st.exchange(soci::use(buyingAsset));
+    st.exchange(soci::use(numOffers));
+
+    {
+        auto timer = mDatabase.getSelectTimer("offer");
+        return loadOffers(prep, offers);
+    }
+}
+
+std::deque<LedgerEntry>::const_iterator
+LedgerTxnRoot::Impl::loadBestOffers(std::deque<LedgerEntry>& offers,
+                                    Asset const& buying, Asset const& selling,
+                                    OfferDescriptor const& worseThan,
+                                    size_t numOffers) const
+{
+    // ManageOffer and related operations won't work correctly with an offerID
+    // equal to or exceeding INT64_MAX, so there is no reason to support it
+    // here. We are far from this limit anyway.
+    if (worseThan.offerID == INT64_MAX)
+    {
+        throw std::runtime_error("maximum offerID encountered");
+    }
+
     // price is an approximation of the actual n/d (truncated math, 15 digits)
     // ordering by offerid gives precendence to older offers for fairness
-    sql += " ORDER BY price, offerid LIMIT :n OFFSET :o";
+    std::string sql =
```

Comment:
nit: don't use `SELECT *` (use explicit columns)

Reply:
Good idea. Changed the query to only use `SELECT *` when querying a table that is part of the `WITH` (meaning `r1` and `r2`) because at that point we already have explicit columns. This matches what we do in other queries.